includes/class-wc-dintero-hp-checkout.php

```diff
@@ -2789,134 +2729,159 @@ public function process_refund( $order_id, $amount = null, $reason = '' ) {
 	 * Check if payment capture is possible when the order is changed from on-hold to complete or processing
 	 *
 	 * @param int $order_id Order ID.
+	 * @return bool
 	 */
-	private function check_capture( $order_id ) {
+	private function check_capture( $order_id )
+	{
+
+		/** @var WC_Order $order */
 		$order = wc_get_order( $order_id );
-		if ( ! empty( $order ) &&
-			 $order instanceof WC_Order &&
-			 $order->get_transaction_id() &&
-			 'dintero-hp' === $order->get_payment_method() ) {
+		if (!$order || !($order instanceof WC_Order)) {
+			return false;
+		}
 
-			$transaction_id = $order->get_transaction_id();
-			$transaction    = $this->get_transaction( $transaction_id );
-			if ( ! array_key_exists( 'merchant_reference', $transaction ) ) {
-				return false;
-			}
-			$transaction_order_id = absint( strval( $transaction['merchant_reference'] ) );
-			if ( $transaction_order_id === $order_id ) {
-				$this->capture( $order, $transaction );
-			}
+		$transaction_id = $order->get_transaction_id();
+		if (empty($transaction_id) || 'dintero-hp' !== $order->get_payment_method()) {
+			return false;
 		}
+
+		$transaction = $this->get_transaction( $transaction_id );
+
+		if ($error = $this->extract('error', $transaction)) {
+			$order->add_order_note(__('Could not capture transaction: ') . $error['message']);
+			$order->save_meta_data();
+			return false;
+		}
+
+		$merchant_reference_id = absint(strval($this->extract(
+			'merchant_reference',
+			$transaction,
+			$this->extract('merchant_reference_2', $transaction)
+		)));
+
+		if (!$merchant_reference_id || $merchant_reference_id !== $order_id ) {
+			$order->add_order_note(__('Could not capture transaction: Merchant reference id not found'));
+			$order->save_meta_data();
+			return false;
+		}
+
+		$this->capture($order, $transaction);
+		return true;
+	}
+
+	/**
+	 * Extracting value from an array
+	 *
+	 * @param string $key
+	 * @param array $data
+	 * @param null $default
+	 * @return mixed|null
+	 */
+	private function extract($key, $data, $default = null)
+	{
+		return !empty($data[$key]) ? $data[$key] : $default;
 	}
 
 	/**
 	 * Capture Payment.
 	 */
-	public function capture( $order, $transaction = null ) {
-		if ( ! empty( $order ) &&
-			 $order instanceof WC_Order &&
-			 $order->get_transaction_id() ) {
+	public function capture( $order, $transaction = null )
```

Comment:
@chaikovskyia What is the difference between this capture-function and the one here: https://github.com/Dintero/Dintero.Checkout.WooCommerce/blob/0cb55c0b57f427fbd95246aec586939b25685bde/includes/class-wc-gateway-dintero-hp.php#L1400

Reply:
@andersem , most likely that is a duplication, but need to check in more details and move it to a separate class.